core/src/main/scala/cats/collections/TreeList.scala

```diff
@@ -0,0 +1,867 @@
+  /**
+   * Convert to a scala standard List
+   */
+  def toList: List[A]
+
+  /**
+   * Convert to a scala standard list, but reversed
+   */
+  def toListReverse: List[A]
+
+  /**
+   * return the right most full binary tree on the right, the rest on left
+   * val (l, r) = items.split
+   * assert((l ++ r) == items)
+   */
+  def split: (TreeList[A], TreeList[A])
+
+  /*
+   * The following methods do not have an optimized
+   * implementation and are expressed in terms of the above
+   */
+  final def ::[A1 >: A](a1: A1): TreeList[A1] = prepend(a1)
+
+  // This is a test method to ensure the invariant on depth is correct
+  private[collections] def maxDepth: Int
+
+  override def toString: String = {
+    val strb = new java.lang.StringBuilder
+    strb.append("TreeList(")
+    def loop(first: Boolean, l: TreeList[A]): Unit =
+      l.uncons match {
+        case None => ()
+        case Some((h, t)) =>
+          if (!first) strb.append(", ")
+          strb.append(h.toString)
+          loop(false, t)
+      }
+
+    loop(true, this)
+    strb.append(")")
+    strb.toString
+  }
+
+  /**
+   * Concatenate two TreeLists. This requires doing as
+   * much work as this.size
+   */
+  final def ++[A1 >: A](that: TreeList[A1]): TreeList[A1] = {
+    @tailrec
+    def loop(ls: List[A], that: TreeList[A1]): TreeList[A1] =
+      ls match {
+        case Nil => that
+        case h :: tail => loop(tail, h :: that)
+      }
+    loop(toListReverse, that)
+  }
+
+  /**
+   * keep the elements that match a predicate
+   */
+  final def filter(fn: A => Boolean): TreeList[A] = {
+    val as = toIterator
+    var resList = List.empty[A]
+    var changed = false
+    while(as.hasNext) {
+      val a = as.next()
+      if (fn(a)) {
+        resList = a :: resList
+      } else { changed = true }
+    }
+    if (changed) TreeList.fromListReverse(resList)
+    else this
+  }
+
+  final def filterNot(fn: A => Boolean): TreeList[A] = {
+    val as = toIterator
+    var resList = List.empty[A]
+    var changed = false
+    while(as.hasNext) {
+      val a = as.next()
+      if (!fn(a)) {
+        resList = a :: resList
+      } else { changed = true }
+    }
+    if (changed) TreeList.fromListReverse(resList)
+    else this
+  }
+
+  /**
+   * Standard flatMap on a List type.
+   */
+  final def flatMap[B](fn: A => TreeList[B]): TreeList[B] = {
+    @tailrec
+    def loop(rev: List[A], acc: TreeList[B]): TreeList[B] =
+      rev match {
+        case Nil => acc
+        case h :: tail =>
+          loop(tail, fn(h) ++ acc)
+      }
+    loop(toListReverse, TreeList.Empty)
+  }
+
+  /**
+   * O(N) reversal of the treeList
+   */
+  final def reverse: TreeList[A] = {
+    val revTrees = toIterator
+    var res: TreeList[A] = TreeList.Empty
+    while(revTrees.hasNext) {
+      res = revTrees.next() :: res
+    }
+    res
+  }
+
+  /**
+   * Take the first n things off the list. O(n)
+   */
+  final def take(n: Long): TreeList[A] = {
+    val takeIt = toIterator
```

Comment:
if n > size we could just return this, 
perhaps a micro-optimization

Reply:
since computing size itself is `O(log N)`, I'm going to leave this off for now until we are more sure we have evidence it is faster. If people usually take small numbers of things, it could be a deoptimization on large lists.

Reply:
so my understanding is that List will have at most 3 tree elements and for Root size is constant call (1) and for Balanced size is cached in already calculated field
so calculating TreeLIst size will take 3 iterations (at most) and adding 3 ints which looks O(1) to me

i guess i'm missing something, sorry for taking your time on that

Reply:
Like having a TreeLIst of 101 elements
```val tl = TreeList.fromList((0 to 100).toList)```

I see this

![image](https://user-images.githubusercontent.com/124249/50720563-787e2480-107d-11e9-9437-5f6f249e7685.png)

Reply:
I don't think it is true that we have O(1) items in the list. I think we have O(log N) items. Since each tree is fully balanced, the size is `2^n - 1` for some n. So, I think for various sizes you have:
```
0 -> []
1 -> [1]
2 -> [1, 1]
3 -> [3]
4 -> [1, 3]
5 -> [1, 1, 3]
6 -> [3, 3]
7 -> [7]
8 -> [1, 7]
...
12 -> [1, 1, 3, 7]
...
```
Note, 12 has 4 items in the list.

Reply:
I think i've got it myself
it's not 3, it's indeed long N